p2p/message.go

```diff
@@ -131,44 +135,181 @@ func (bm *BlockRequestMessage) String() string {
+		if len(num) < 8 {
+			num = common.AppendZeroes(num, 8)
+		}
+		encMsg = append(encMsg, num...)
+	} else {
+		encMsg = append(encMsg, bm.StartingBlock...)
 	}
 
-	if bm.EndBlockHash != [32]byte{} {
-		encMsg = append(encMsg, bm.EndBlockHash.ToBytes()...)
+	if bm.EndBlockHash == nil || bm.EndBlockHash[0] == 0 {
+		encMsg = append(encMsg, []byte{0, 0}...)
 	} else {
-		encMsg = append(encMsg, 0)
+		encMsg = append(encMsg, bm.EndBlockHash...)
 	}
 
 	encMsg = append(encMsg, bm.Direction)
 
-	if bm.Max != 0 {
-		encMax := make([]byte, 4)
-		binary.LittleEndian.PutUint32(encMax, bm.Max)
-		encMsg = append(encMsg, encMax...)
+	if bm.Max == nil || bm.Max[0] == 0 {
+		encMsg = append(encMsg, []byte{0, 0}...)
 	} else {
-		encMsg = append(encMsg, 0)
+		encMax := make([]byte, 5)
+		copy(encMax, bm.Max)
+		encMsg = append(encMsg, encMax...)
 	}
 
 	return encMsg, nil
 }
 
 // Decodes the message into a BlockRequestMessage, it assumes the type byte has been removed
-func (bm *BlockRequestMessage) Decode(msg []byte) error {
+func (bm *BlockRequestMessage) Decode(r io.Reader) error {
+	var err error
+	bm.Id, err = readUint64(r)
+	if err != nil {
+		return err
+	}
+
+	bm.RequestedData, err = readByte(r)
+	if err != nil {
+		return err
+	}
+
+	startingBlockType, err := readByte(r)
+	if err != nil {
+		return err
+	}
+
+	if startingBlockType == 0 {
+		hash := make([]byte, 32)
+		_, err = r.Read(hash)
+		if err != nil {
+			return err
+		}
+		bm.StartingBlock = append([]byte{startingBlockType}, hash...)
+	} else {
+		num := make([]byte, 8)
+		_, err = r.Read(num)
+		if err != nil {
+			return err
+		}
+		bm.StartingBlock = append([]byte{startingBlockType}, num...)
+	}
+
+	endBlockHashExists, err := readByte(r)
+	if err != nil {
+		return err
+	}
+
+	// if endBlockHash was None, then just set Direction and Max
+	if endBlockHashExists == 0 {
+		bm.EndBlockHash = []byte{0, 0}
+	} else {
+		endBlockHash := make([]byte, 32)
+		_, err = r.Read(endBlockHash)
+		if err != nil {
+			return err
+		}
+		bm.EndBlockHash = append([]byte{1}, endBlockHash...)
+	}
+
+	dir, err := readByte(r)
+	if err != nil {
+		return err
+	}
+
+	bm.Direction = dir
+
+	maxExists, err := readByte(r)
+	if err != nil {
+		return err
+	}
+
+	if maxExists == 0 {
+		bm.Max = []byte{0, 0}
+	} else {
+		max := make([]byte, 4)
+		_, err = r.Read(max)
+		if err != nil {
+			return err
+		}
+		bm.Max = append([]byte{1}, max...)
+	}
+
+	return nil
+}
+
+type BlockResponseMessage struct {
+	Id   uint64
+	Data []byte // TODO: change this to BlockData type
+}
+
+func (bm *BlockResponseMessage) String() string {
+	return fmt.Sprintf("BlockResponseMessage Id=%d Data=%x", bm.Id, bm.Data)
+}
+
+func (bm *BlockResponseMessage) Encode() ([]byte, error) {
+	encMsg := []byte{BlockResponseMsgType}
+
+	encId := make([]byte, 8)
+	binary.LittleEndian.PutUint64(encId, bm.Id)
+	encMsg = append(encMsg, encId...)
+
+	return append(encMsg, bm.Data...), nil
+}
+
+func (bm *BlockResponseMessage) Decode(r io.Reader) error {
```

Comment:
Please write a test for `(bm *BlockResponseMessage) Decode(r io.Reader) error` and comment

Reply:
done